core/src/main/java/org/infinispan/remoting/transport/jgroups/CommandAwareRpcDispatcher.java

```diff
@@ -272,6 +289,78 @@ private SingleResponseFuture processSingleCall(ReplicableCommand command, long t
       return retval;
    }
 
+   private void processCallsStaggered(ReplicableCommand command, RspFilter filter, List<Address> dests,
+                                      ResponseMode mode, DeliverOrder deliverOrder, Marshaller marshaller,
+                                      CompletableFuture<RspList<Response>> theFuture, int destIndex,
+                                      long deadline, RspList<Response> rsps)
+         throws Exception {
+      if (destIndex == dests.size())
+         return;
+
+      CompletableFuture<Void> triggerNextFuture = new CompletableFuture<>();
+      CompletableFuture<Rsp<Response>> subFuture =
+            processSingleCall(command, -1, dests.get(destIndex), mode, deliverOrder, marshaller);
+      if (subFuture != null) {
+         subFuture.whenComplete((rsp, throwable) -> {
+            if (throwable != null) {
+               // We should never get here, any remote exception will be in the Rsp
+               theFuture.completeExceptionally(throwable);
+            }
+            Rsp<Response> futureRsp = rsps.get(rsp.getSender());
+            if (rsp.hasException()) {
+               futureRsp.setException(rsp.getException());
+            } else {
+               futureRsp.setValue(rsp.getValue());
+            }
+            if (filter.isAcceptable(rsp.getValue(), rsp.getSender())) {
+               // We got an acceptable response
+               theFuture.complete(rsps);
+            } else {
+               boolean missingResponses = false;
+               for (Rsp<Response> rsp1 : rsps) {
+                  if (!rsp1.wasReceived()) {
+                     missingResponses = true;
+                     break;
+                  }
+               }
+               if (!missingResponses) {
+                  // This was the last response, need to complete the future
+                  theFuture.complete(rsps);
+               } else {
+                  // The response was not acceptable, complete the timeout future to start the next request
+                  triggerNextFuture.complete(null);
+               }
+            }
+         });
+         if (!subFuture.isDone()) {
+            long now = System.nanoTime();
+            long delay = deadline - now;
+            if (destIndex < dests.size() - 1) {
+               delay = Math.min(STAGGER_DELAY_NANOS, delay);
+            } // else it is the last request, so schedule the cancelation at deadline
+            timeoutExecutor.schedule(() -> triggerNextFuture.complete(null),
```

Comment:
Do you think we need to retain the future returned here, in case if the stagger amount is much higher than a regular response?  In that case we could build up these timer tasks.